packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -263,13 +269,15 @@ export async function accumulateMetadata(
 
     parentPromise = parentPromise.then((resolved) => {
       return layerMetadataPromise.then((metadata) => {
+        resolved = deepClone(resolved)
         merge(resolved, metadata, staticFilesMetadata, {
           title: resolved.title?.template || null,
           openGraph: resolved.openGraph?.title?.template || null,
           twitter: resolved.twitter?.title?.template || null,
         })
-
-        return resolved
+        return process.env.NODE_ENV === 'development'
+          ? Object.freeze(resolved)
+          : resolved
```

Comment:
I would move the clone the dev branch. the DCE would probably take care of the noop clone in prod but better to be sure by structuring the code so it doesn't require extra smart DCE. This only works if you change the above merging to target `resolvedMetadata` as described above

```
return process.env.NODE_ENV === 'development'
  ? Object.freeze(deepClone(resolved))
  : resolved
```